modules/tinymce/src/core/main/ts/keyboard/Nbsps.ts

```diff
@@ -149,22 +162,37 @@ const normalizeNbspInMiddleOfTextNode = (node: Text): boolean => {
   }
 };
 
-const normalizeNbspAtEnd = (root: SugarElement<Node>, node: Text): boolean => {
+const normalizeNbspAtEnd = (root: SugarElement<Node>, node: Text, makeNbsp: boolean): boolean => {
   const text = node.data;
   const lastPos = CaretPosition(node, text.length - 1);
-  if (isNbspAt(text, text.length - 1) && !needsToBeNbsp(root, lastPos)) {
+  if (isNbspAt(text, text.length - 1) && !needsToBeNbsp(root, lastPos) && !makeNbsp) {
     node.data = text.slice(0, -1) + ' ';
     return true;
+  } else if (makeNbsp && isWhiteSpaceAt(text, text.length - 1) && needsToBeNbspRight(root, lastPos)) {
+    node.data = text.slice(0, -1) + Unicode.nbsp;
+    return true;
   } else {
     return false;
   }
 };
 
-const normalizeNbsps = (root: SugarElement<Node>, pos: CaretPosition): Optional<CaretPosition> => Optional.some(pos).filter(hasNbsp).bind((pos) => {
-  const container = pos.container() as Text;
-  const normalized = normalizeNbspAtStart(root, container) || normalizeNbspInMiddleOfTextNode(container) || normalizeNbspAtEnd(root, container);
-  return normalized ? Optional.some(pos) : Optional.none();
-});
+const normalizeNbsps = (root: SugarElement<Node>, pos: CaretPosition): Optional<CaretPosition> =>
+  Optional.some(pos).filter(hasNbsp).fold(
```

Comment:
I agree, there's a place and time for optionals but doing _anything_ with an `Optional.some()` other than returning it or passing it to a function is generally incorrect.

I'm also dubious about this fold implementation, it's using `pos` from the function argument in the `none` branch which is extremely confusing. Once it's unpacked into a basic if/else structure it might be more obvious where to combine branches.